src/coreclr/jit/async.cpp

```diff
@@ -37,6 +45,240 @@
 #include "jitstd/algorithm.h"
 #include "async.h"
 
+//------------------------------------------------------------------------
+// Compiler::SaveAsyncContexts:
+//   Insert code to save and restore ExecutionContext around async call sites.
+//
+// Returns:
+//   Suitable phase status.
+//
+// Remarks:
+//   Runs early, after import but before inlining. Thus RET_EXPRs may be
+//   present, and async calls may later be inlined.
+//
+PhaseStatus Compiler::SaveAsyncContexts()
+{
+    if (!compMustSaveAsyncContexts)
+    {
+        JITDUMP("No async calls where async context capture/restore is necessary\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
```

Comment:
We often put debug checks to verify that these early outs are correct.

So no calls if non expected, and no non-top level calls, etc. Worth doing?